core/drivers/gpio/gpio.c

```diff
@@ -29,43 +32,72 @@ TEE_Result gpio_dt_alloc_pin(struct dt_pargs *pargs, struct gpio **out_gpio)
 	return TEE_SUCCESS;
 }
 
-static char *gpio_get_dt_prop_name(const char *gpio_name)
-{
-	int ret = 0;
-	char *prop_name = NULL;
-	size_t max_len = strlen(gpio_name) + strlen("-gpios") + 1;
-
-	prop_name = calloc(1, max_len);
-	if (!prop_name)
-		return NULL;
-
-	ret = snprintf(prop_name, max_len, "%s-gpios", gpio_name);
-	if (ret < 0 || (size_t)ret >= max_len) {
-		free(prop_name);
-		return NULL;
-	}
-
-	return prop_name;
-}
-
 TEE_Result gpio_dt_get_by_index(const void *fdt, int nodeoffset,
 				unsigned int index, const char *gpio_name,
 				struct gpio **gpio)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	char *prop_name = NULL;
+	char prop_name[32]; /* 32 is max size of property name in DT */
 	void *out_gpio = NULL;
+	unsigned int i = 0;
 
-	prop_name = gpio_get_dt_prop_name(gpio_name);
-	if (!prop_name)
-		return TEE_ERROR_OUT_OF_MEMORY;
+	/* Try GPIO properties "foo-gpios" and "foo-gpio" */
+	for (i = 0; i < ARRAY_SIZE(gpio_suffixes); i++) {
+		if (gpio_name)
+			snprintf(prop_name, sizeof(prop_name), "%s-%s",
+				 gpio_name, gpio_suffixes[i]);
+		else
+			snprintf(prop_name, sizeof(prop_name), "%s",
+				 gpio_suffixes[i]);
+
+		res = dt_driver_device_from_node_idx_prop(prop_name, fdt,
+							  nodeoffset,
+							  index, DT_DRIVER_GPIO,
+							  &out_gpio);
+
+		if (res != TEE_ERROR_ITEM_NOT_FOUND)
+			break;
+	}
 
-	res = dt_driver_device_from_node_idx_prop(prop_name, fdt, nodeoffset,
-						  index, DT_DRIVER_GPIO,
-						  &out_gpio);
-	free(prop_name);
 	if (!res)
 		*gpio = out_gpio;
 
 	return res;
 }
+
+TEE_Result gpio_configure(struct gpio *gpio, enum gpio_flags flags)
+{
+	enum gpio_level value = GPIO_LEVEL_LOW;
+
+	assert(gpio && gpio->chip && gpio->chip->ops);
+
+	/* Configure GPIO with DT flags */
+	if (gpio && gpio->chip->ops->configure)
+		gpio->chip->ops->configure(gpio->chip, gpio);
+
+	/* Process requester flags */
+	if (flags & GPIO_FLAGS_BIT_DIR_SET) {
+		if (flags & GPIO_FLAGS_BIT_DIR_OUT) {
+			if (flags & GPIO_FLAGS_BIT_DIR_VAL)
+				value = GPIO_LEVEL_HIGH;
+			gpio_set_value(gpio, value);
+			gpio_set_direction(gpio, GPIO_DIR_OUT);
+		} else {
+			gpio_set_direction(gpio, GPIO_DIR_IN);
+		}
+	}
+
+	return TEE_SUCCESS;
```

Comment:
If always successful, return value is not needed.
That said, I think it would be better to allow `configure` ops handler to return a status.